src/ring_theory/tensor_product.lean

```diff
@@ -1,68 +1,315 @@
 /-
 Copyright (c) 2020 Scott Morrison. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Scott Morrison
+Authors: Scott Morrison, Johan Commelin
 -/
 
 import linear_algebra.tensor_product
-import algebra.algebra.basic
+import algebra.algebra.tower
 
 universes u v₁ v₂ v₃ v₄
 
-
 /-!
 # The tensor product of R-algebras
 
-We construct the R-algebra structure on `A ⊗[R] B`, when `A` and `B` are both `R`-algebras,
-and provide the structure isomorphisms
+Let `R` be a (semi)ring and `A` an `R`-algebra.
+In this file we:
 
-* `R ⊗[R] A ≃ₐ[R] A`
-* `A ⊗[R] R ≃ₐ[R] A`
-* `A ⊗[R] B ≃ₐ[R] B ⊗[R] A`
+- Define the `A`-module structure on `A ⊗ M`, for an `R`-module `M`.
+- Define the `R`-algebra structure on `A ⊗ B`, for another `R`-algebra `B`.
+  and provide the structure isomorphisms
+  * `R ⊗[R] A ≃ₐ[R] A`
+  * `A ⊗[R] R ≃ₐ[R] A`
+  * `A ⊗[R] B ≃ₐ[R] B ⊗[R] A`
+  * `((A ⊗[R] B) ⊗[R] C) ≃ₐ[R] (A ⊗[R] (B ⊗[R] C))`
 
-The code for
-* `((A ⊗[R] B) ⊗[R] C) ≃ₐ[R] (A ⊗[R] (B ⊗[R] C))`
-is written and compiles, but takes longer than the `-T100000` time limit,
-so is currently commented out.
--/
+## Main declaration
 
-namespace algebra
+- `linear_map.base_change A f` is the `A`-linear map `A ⊗ f`, for an `R`-linear map `f`.
+
+-/
 
 open_locale tensor_product
 open tensor_product
 
 namespace tensor_product
 
+variables {R A M N P : Type*}
+
+section restrict_scalars
+variables [comm_semiring R] [comm_semiring A] [algebra R A]
+variables [add_comm_monoid M] [semimodule A M]
+variables [add_comm_monoid N] [semimodule A N]
+
+variables (R A M N)
+
+@[priority 500]
+instance restrict_scalars.semimodule : semimodule R (M ⊗[A] N) :=
```

Comment:
Doesn't this already exist as `tensor_product.semimodule'`?

Reply:
What you say is almost true but a type synonym gets in the way. More precisely, the following fails with a complaint about not being able to find `[module R M]` and `[module R N]`:
```lean
def restrict_scalars.module : module R (M ⊗[A] N) :=
tensor_product.module'
```
but the following works:
```lean
def restrict_scalars.module : module R ((restrict_scalars R A M) ⊗[A] (restrict_scalars R A N)) :=
tensor_product.module'
```
I'm still trying to decide what the best setup is here.

(I realise this is out of date; these remarks are just to leave me with a clear conscience after I hit the resolve button.)